pgrx/src/pg_catalog.rs

```diff
@@ -0,0 +1,1284 @@
+    catalog (pg_am) {
+        /// Row identifier
+        (oid, Oid, get_struct)
+        /// Name of the access method
+        (amname, &CStr, get_struct)
+        /// OID of a handler function that is responsible for supplying information about the access method
+        (amhandler, Regproc, get_struct)
+        (amtype, PgAmAmtype, character {
+            /// t = table (including materialized views)
+            (Table, b't')
+            /// i = index
+            (Index, b'i')
+        }, get_struct)
+    }
+}
+
+define_catalog! {
+    /// The catalog pg_amop stores information about operators associated with access method operator families. There is one row for each operator that is a member of an operator family. A family member can be either a search operator or an ordering operator. An operator can appear in more than one family, but cannot appear in more than one search position nor more than one ordering position within a family. (It is allowed, though unlikely, for an operator to be used for both search and ordering purposes.)
+    catalog (pg_amop) {
+        /// Row identifier
+        (oid, Oid, get_struct)
+        /// The operator family this entry is for
+        (amopfamily, Oid, get_struct)
+        /// Left-hand input data type of operator
+        (amoplefttype, Oid, get_struct)
+        /// Right-hand input data type of operator
+        (amoprighttype, Oid, get_struct)
+        /// Operator strategy number
+        (amopstrategy, i16, get_struct)
+        /// Operator purpose
+        (amoppurpose, PgAmopAmoppurpose, character {
+            /// s for search
+            (Search, b's')
+            /// o for ordering
+            (Order, b'o')
+        }, get_struct)
+        /// OID of the operator
+        (amopopr, Oid, get_struct)
+        /// Index access method operator family is for
+        (amopmethod, Oid, get_struct)
+        /// The B-tree operator family this entry sorts according to, if an ordering operator; zero if a search operator
+        (amopsortfamily, Oid, get_struct)
+    }
+}
+
+define_catalog! {
+    /// The catalog pg_amproc stores information about support functions associated with access method operator families. There is one row for each support function belonging to an operator family.
+    catalog (pg_amproc) {
+        /// Row identifier
+        (oid, Oid, get_struct)
+        /// The operator family this entry is for
+        (amprocfamily, Oid, get_struct)
+        /// Left-hand input data type of associated operator
+        (amproclefttype, Oid, get_struct)
+        /// Right-hand input data type of associated operator
+        (amprocrighttype, Oid, get_struct)
+        /// Support function number
+        (amprocnum, i16, get_struct)
+        /// OID of the function
+        (amproc, Regproc, get_struct)
+    }
+}
+
+define_catalog! {
```

Comment:
Just picking one of these at random...

You hand-wrote all of these `define_catalog!()` declarations?  I don't see that as sustainable.  Is there any way we can generate this, via `build.rs` from the Postgres sources?

Even if that meant we'd need our own equivalent of `genbki.pl` (in Rust, obviously), that will be infinitely better than trying to manually keep these up-to-date.  Postgres does change catalogs from time to time.

Reply:
All declarations are hand-written. If a new column is added in a newer version of PostgreSQL, who uses it could add it again, so I thought it's not a problem.

It looks that this process could be done by parsing `catalog/pg_*.h` and `catalogs.sgml`. I will try it later.

Reply:
> who uses it could add it again

The problem with that is it's not guaranteed anyone would notice, which might lead to miscompilations or runtime UB or who-knows-what.

In `build.rs`, we ultimately have the struct definitions as provided by bindgen.  Maybe those can be traversed to generate the remaining code?

Ultimately, this would get us 100% coverage of all the catalogs.

Reply:
That does seem like it would be best, although it might be a bit toilsome to implement the first time and occasionally to keep up to date.